src/create-infura-middleware.ts

```diff
@@ -0,0 +1,218 @@
+        // otherwise, ignore error and retry again after timeout
+        await timeout(1000);
+      }
+    }
+    // request was handled correctly, end
+  });
+}
+
+/**
+ * Makes a request to Infura, updating the given response object if the response
+ * has a "successful" status code or throwing an error otherwise.
+ *
+ * @param network - A network that Infura supports; plugs into
+ * `https://${network}.infura.io`.
+ * @param projectId - The Infura project id.
+ * @param extraHeaders - Extra headers that will be used to make the request.
+ * @param req - The original request object obtained via the middleware stack.
+ * @param res - The original response object obtained via the middleware stack.
+ * @param source - A descriptor for the entity making the request;
+ * tracked by Infura for analytics purposes.
+ * @throws an error with a detailed message if the HTTP status code is anywhere
+ * outside 2xx, and especially if it is 405, 429, 503, or 504.
+ */
+async function performFetch(
+  network: InfuraJsonRpcSupportedNetwork,
+  projectId: string,
+  extraHeaders: RequestHeaders,
+  req: ExtendedJsonRpcRequest<unknown>,
+  res: PendingJsonRpcResponse<unknown>,
+  source: string | undefined,
+): Promise<void> {
+  const { fetchUrl, fetchParams } = fetchConfigFromReq({
+    network,
+    projectId,
+    extraHeaders,
+    req,
+    source,
+  });
+  const response = await fetch(fetchUrl, fetchParams);
+  const rawData = await response.text();
+  // handle errors
+  if (!response.ok) {
+    switch (response.status) {
+      case 405:
+        throw ethErrors.rpc.methodNotFound();
+
+      case 429:
+        throw createRatelimitError();
+
+      case 503:
+      case 504:
+        throw createTimeoutError();
+
+      default:
+        throw createInternalError(rawData);
+    }
+  }
+
+  // special case for now
+  if (req.method === 'eth_getBlockByNumber' && rawData === 'Not Found') {
+    res.result = null;
+    return;
+  }
+
+  // parse JSON
+  const data = JSON.parse(rawData);
+
+  // finally return result
+  res.result = data.result;
+  res.error = data.error;
+}
+
+/**
+ * Builds a JSON-RPC 2.0 internal error object describing a rate-limiting
+ * error.
+ *
+ * @returns The error object.
+ */
+function createRatelimitError(): EthereumRpcError<unknown> {
+  const msg = `Request is being rate limited.`;
+  return createInternalError(msg);
+}
+
+/**
+ * Builds a JSON-RPC 2.0 internal error object describing a timeout error.
+ *
+ * @returns The error object.
+ */
+function createTimeoutError(): EthereumRpcError<unknown> {
+  let msg = `Gateway timeout. The request took too long to process. `;
+  msg += `This can happen when querying logs over too wide a block range.`;
+  return createInternalError(msg);
+}
+
+/**
+ * Builds a JSON-RPC 2.0 internal error object.
+ *
+ * @param msg - The message.
+ * @returns The error object.
+ */
+function createInternalError(msg: string): EthereumRpcError<unknown> {
+  return ethErrors.rpc.internal(msg);
+}
```

Comment:
We can be more specific here, we _know_ that `data` is `undefined`.

```suggestion
function createRatelimitError(): EthereumRpcError<undefined> {
  const msg = `Request is being rate limited.`;
  return createInternalError(msg);
}

/**
 * Builds a JSON-RPC 2.0 internal error object describing a timeout error.
 *
 * @returns The error object.
 */
function createTimeoutError(): EthereumRpcError<undefined> {
  let msg = `Gateway timeout. The request took too long to process. `;
  msg += `This can happen when querying logs over too wide a block range.`;
  return createInternalError(msg);
}

/**
 * Builds a JSON-RPC 2.0 internal error object.
 *
 * @param msg - The message.
 * @returns The error object.
 */
function createInternalError(msg: string): EthereumRpcError<undefined> {
  return ethErrors.rpc.internal(msg);
}
```